### PR TITLE
Laboratory work #3, Ekaterina Khomutova - 22FPL2

### PR DESCRIPTION
**Pull request by EkaterinaKhomutova**

### CODE REVIEW

**Review by marina-kaz — COMMENTED** [review]
Strong start, keep going!

No pressure, but I have high hopes!

File: lab_3_generate_by_ngrams/main.py [hunk 1]
```diff
@@ -41,6 +45,24 @@ def _tokenize(self, text: str) -> Optional[tuple[str, ...]]:
...
+        if text[-1] in '.,:;?! ':
+            flag = True
+        for word in words:
+            tokenized_word = [token.lower() for token in word if token.isalpha()]
+            if tokenized_word:
+                tokens.extend(tokenized_word)
+                tokens.append(self._end_of_word_token)
+        if tokens == []:
+            return None
+        if not flag:
```
[the remark this answers is not in the document]
Comment (EkaterinaKhomutova) [reply lab_3_generate_by_ngrams/main.py:62]: Done (both)

File: lab_3_generate_by_ngrams/main.py [hunk 2]
```diff
@@ -76,6 +107,13 @@ def get_token(self, element_id: int) -> Optional[str]:
 
         In case of corrupt input arguments or arguments not included in storage, None is returned
         """
+        if not isinstance(element_id, int):
+            return None
+        if len(self._storage) <= element_id:
```
[the remark this answers is not in the document]
Comment (EkaterinaKhomutova) [reply lab_3_generate_by_ngrams/main.py:112]: How can I do it more starightforward?

File: lab_3_generate_by_ngrams/main.py [hunk 3]
```diff
@@ -76,6 +107,13 @@ def get_token(self, element_id: int) -> Optional[str]:
...
+            return None
+
+        return list(self._storage.keys())[element_id]
```
[the remark this answers is not in the document]
Comment (EkaterinaKhomutova) [reply lab_3_generate_by_ngrams/main.py:115]: fixed (I guess)

File: lab_3_generate_by_ngrams/main.py [hunk 4]
```diff
@@ -93,6 +131,25 @@ def encode(self, text: str) -> Optional[tuple[int, ...]]:
         In case of corrupt input arguments, None is returned.
         In case any of methods used return None, None is returned.
         """
+        if (
+            not isinstance(text, str) or
+            text == ''
```
[the remark this answers is not in the document]
Comment (EkaterinaKhomutova) [reply lab_3_generate_by_ngrams/main.py:136]: I changed it to 'not text'

File: lab_3_generate_by_ngrams/main.py [hunk 5]
```diff
@@ -121,6 +186,20 @@ def decode(self, encoded_corpus: tuple[int, ...]) -> Optional[str]:
...
+        if (
+                not isinstance(encoded_corpus, tuple) or
+                encoded_corpus == ()
+        ):
+            return None
+        decoded_corpus = self._decode(encoded_corpus)
+        if not decoded_corpus:
+            return None
+        decoded_text = self._postprocess_decoded_text(decoded_corpus)
+        if not decoded_text:
```
[the remark this answers is not in the document]
Comment (EkaterinaKhomutova) [reply lab_3_generate_by_ngrams/main.py:198]: I got it. Fixed

File: lab_3_generate_by_ngrams/main.py [hunk 6]
```diff
@@ -5,6 +5,7 @@
 """
 # pylint:disable=too-few-public-methods
 from typing import Optional
+from string import *
```
Comment (marina-kaz) [inline lab_3_generate_by_ngrams/main.py:9]: Why use a wildcard? Are there that many string fields you use in your code?

File: lab_3_generate_by_ngrams/main.py [hunk 7]
```diff
@@ -41,6 +46,22 @@ def _tokenize(self, text: str) -> Optional[tuple[str, ...]]:
...
+            return None
+        words = text.split()
+        tokens = []
+
+        for word in words:
+            tokenized_word = [token.lower() for token in word if token.isalpha()]
+            if tokenized_word:
+                tokens.extend(tokenized_word)
+                tokens.append(self._end_of_word_token)
+        if tokens == []:
```
Comment (marina-kaz) [inline lab_3_generate_by_ngrams/main.py:59]: Empty sequences are mapped to False automatically

File: lab_3_generate_by_ngrams/main.py [hunk 8]
```diff
@@ -41,6 +46,22 @@ def _tokenize(self, text: str) -> Optional[tuple[str, ...]]:
...
+            return None
+        if text[-1] not in punctuation:
```
Comment (marina-kaz) [inline lab_3_generate_by_ngrams/main.py:61]: what if it's a space character? 👩‍🚀

File: lab_3_generate_by_ngrams/main.py [hunk 9]
```diff
@@ -76,6 +107,15 @@ def get_token(self, element_id: int) -> Optional[str]:
 
         In case of corrupt input arguments or arguments not included in storage, None is returned
         """
+        if (
+            not isinstance(element_id, int) or
+            len(self._storage) <= element_id
+        ):
+            return None
+        token = [i for i in self._storage if self._storage[i] == element_id][0]
```
Comment (marina-kaz) [inline lab_3_generate_by_ngrams/main.py:115]: looks like there must be just one such element, so using list and indexing is a bit extra here

File: lab_3_generate_by_ngrams/main.py [hunk 10]
```diff
@@ -221,6 +346,20 @@ def generate_next_token(self, sequence: tuple[int, ...]) -> Optional[dict]:
...
+            not(isinstance(sequence, tuple) and sequence)
+            or len(sequence) < self._n_gram_size - 1
+        ):
+            return None
+        context = sequence[-(self._n_gram_size-1):]
+        next_token_freq = {}
+        for n_gram, frequency in self._n_gram_frequencies.items():
+            if n_gram[:-1] == context:
+                token = n_gram[-1]
+                if token not in next_token_freq:
```
Comment (marina-kaz) [inline lab_3_generate_by_ngrams/main.py:359]: Could it be?

**Review by marina-kaz — COMMENTED** [review]
Good! Keep going, I think you can nail the next tasks as well!

File: lab_3_generate_by_ngrams/main.py (hunk 6 above)
Comment (EkaterinaKhomutova) [reply lab_3_generate_by_ngrams/main.py:9]: I just thought that it seems nice to write import in such way

File: lab_3_generate_by_ngrams/main.py (hunk 8 above)
Comment (EkaterinaKhomutova) [reply lab_3_generate_by_ngrams/main.py:61]: fixed

File: lab_3_generate_by_ngrams/main.py (hunk 9 above)
Comment (EkaterinaKhomutova) [reply lab_3_generate_by_ngrams/main.py:115]: rewritten by filter()

File: lab_3_generate_by_ngrams/main.py (hunk 10 above)
Comment (EkaterinaKhomutova) [reply lab_3_generate_by_ngrams/main.py:359]: fixed

File: lab_3_generate_by_ngrams/main.py (hunk 6 above)
Comment (marina-kaz) [reply lab_3_generate_by_ngrams/main.py:11]: Certainly a fancy one, but useless here

File: lab_3_generate_by_ngrams/main.py [hunk 11]
```diff
@@ -41,6 +47,21 @@ def _tokenize(self, text: str) -> Optional[tuple[str, ...]]:
...
+        tokens = []
+
+        for word in words:
+            tokenized_word = [token.lower() for token in word if token.isalpha()]
+            if tokenized_word:
+                tokens.extend(tokenized_word)
+                tokens.append(self._end_of_word_token)
+        if not tokens:
+            return None
+        if text[-1] not in string.punctuation and text[-1]!= ' ':
```
Comment (marina-kaz) [inline lab_3_generate_by_ngrams/main.py:62]: what if the last element is `\n`?

File: lab_3_generate_by_ngrams/main.py [hunk 12]
```diff
@@ -129,6 +200,13 @@ def fill_from_ngrams(self, content: dict) -> None:
         Args:
             content (dict): ngrams from external JSON
         """
+        if not(isinstance(content, dict) and content):
+            return None
+
+        n_grams = ''.join([''.join((filter(lambda x: x.isalpha, key))) for key in content['freq']])
```
Comment (marina-kaz) [inline lab_3_generate_by_ngrams/main.py:206]: I believe spaces are to be replaces with eos token, and then we should filter by letters

File: lab_3_generate_by_ngrams/main.py [hunk 13]
```diff
@@ -208,6 +315,22 @@ def build(self) -> int:
...
+            return 1
+        n_grams = self._extract_n_grams(self._encoded_corpus)
+        if not n_grams:
+            return 1
+        for n_gram in n_grams:
+            if n_gram in self._n_gram_frequencies:
+                continue
+            count_current = n_grams.count(n_gram)
+            cut_ngram = n_gram[:self._n_gram_size-1]
+            starts_from_n_gram = len([i for i in n_grams if i[:self._n_gram_size-1] == cut_ngram])
```
Comment (marina-kaz) [inline lab_3_generate_by_ngrams/main.py:330]: is there a simpler way to index the last element?

File: lab_3_generate_by_ngrams/main.py [hunk 14]
```diff
@@ -388,6 +587,36 @@ def run(self, prompt: str, seq_len: int) -> Optional[str]:
...
+        ):
+            return None
+
+        encoded = self._text_processor.encode(prompt)
+        if not encoded:
+            return None
+        candidates = {encoded: 0.0}
+
+        for i in range(seq_len):
+            copy_candidates = candidates.copy()
```
Comment (marina-kaz) [inline lab_3_generate_by_ngrams/main.py:602]: Good, this is what I would like to see in `continue_sequence`

File: lab_3_generate_by_ngrams/main.py [hunk 15]
```diff
@@ -388,6 +587,36 @@ def run(self, prompt: str, seq_len: int) -> Optional[str]:
...
+            for sequence in candidates:
+                next_tokens = self._get_next_token(sequence)
+                if not next_tokens:
+                    return None
+                next_tokens = sorted(next_tokens, key= lambda x: (x[1], x[0]))[:self._beam_width]
+                continued = self.beam_searcher.continue_sequence(
+                    sequence, next_tokens, copy_candidates
+                )
+                if not continued:
+                    return self._text_processor.decode(sorted(candidates, key=lambda x: x[1])[0])
```
Comment (marina-kaz) [inline lab_3_generate_by_ngrams/main.py:612]: consider breaking a cycle instead of duplicating code

**Review by marina-kaz — COMMENTED** [review]
Good progress!

File: lab_3_generate_by_ngrams/main.py (hunk 11 above)
Comment (EkaterinaKhomutova) [reply lab_3_generate_by_ngrams/main.py:62]: fixed

**Pull request closed without merging**